src/lib/Microsoft.Health.Common/EventHub/Listeners/EventHubListener.cs

```diff
@@ -0,0 +1,498 @@
+using Newtonsoft.Json;
+
+namespace Microsoft.Health.Common.EventHubs
+{
+    internal sealed class EventHubListener : IListener, IEventProcessorFactory, IScaleMonitorProvider
+    {
+        private static readonly Dictionary<string, object> EmptyScope = new Dictionary<string, object>();
+        private readonly string _functionId;
+        private readonly string _eventHubName;
+        private readonly string _consumerGroup;
+        private readonly string _connectionString;
+        private readonly string _storageConnectionString;
+        private readonly ITriggeredFunctionExecutor _executor;
+        private readonly EventProcessorHost _eventProcessorHost;
+        private readonly bool _singleDispatch;
+        private readonly EventHubOptions _options;
+        private readonly ILogger _logger;
+        private bool _started;
+
+        private Lazy<EventHubsScaleMonitor> _scaleMonitor;
+
+        public EventHubListener(
+            string functionId,
+            string eventHubName,
+            string consumerGroup,
+            string connectionString,
+            string storageConnectionString,
+            ITriggeredFunctionExecutor executor,
+            EventProcessorHost eventProcessorHost,
+            bool singleDispatch,
+            EventHubOptions options,
+            ILogger logger,
+            CloudBlobContainer blobContainer = null)
+        {
+            _functionId = functionId;
+            _eventHubName = eventHubName;
+            _consumerGroup = consumerGroup;
+            _connectionString = connectionString;
+            _storageConnectionString = storageConnectionString;
+            _executor = executor;
+            _eventProcessorHost = eventProcessorHost;
+            _singleDispatch = singleDispatch;
+            _options = options;
+            _logger = logger;
+            _scaleMonitor = new Lazy<EventHubsScaleMonitor>(() => new EventHubsScaleMonitor(_functionId, _eventHubName, _consumerGroup, _connectionString, _storageConnectionString, _logger, blobContainer));
+        }
+
+        void IListener.Cancel()
+        {
+            StopAsync(CancellationToken.None).Wait();
+        }
+
+        void IDisposable.Dispose()
+        {
+        }
+
+        public async Task StartAsync(CancellationToken cancellationToken)
+        {
+            await _eventProcessorHost.RegisterEventProcessorFactoryAsync(this, _options.EventProcessorOptions);
+            _started = true;
+        }
+
+        public async Task StopAsync(CancellationToken cancellationToken)
+        {
+            if (_started)
+            {
+                await _eventProcessorHost.UnregisterEventProcessorAsync();
+            }
+
+            _started = false;
+        }
+
+        IEventProcessor IEventProcessorFactory.CreateEventProcessor(PartitionContext context)
+        {
+            return new EventProcessor(_options, _executor, _logger, _singleDispatch);
+        }
+
+        public IScaleMonitor GetMonitor()

```

Comment:
How would we auto-scale this new eventhubtrigger/eventhublistener across azure functions?

Reply:
This is handled by the hosting infrastructure.  If we omit the same metrics as the current EventHubTrigger we should be able to leverage the same built in AKS scaling.  The scaling itself isn't handled in the function, just the telemetry collected.